### PR TITLE
feat(sqlalchemy): Add Support for externalAuthentication

### LINKED ISSUE #343: SQLAlchemy support for externalAuthentication
### Describe the feature

It'd be great to have SQLAlchemy support for the externalAuthentication query parameter since there are a few libraries that use SQLAlchemy underneath and sometimes passing a Python Class through the connection_args is not possible

### Describe alternatives you've considered

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by IceS2**
## Description
Adds support for the 'externalAuthentication' Trino query URL parameter.

If 'externalAuthentication' is passed on the query arguments:
- Sets 'http_scheme' to 'https' 
- Sets 'auth' to OAuth2Authentication()

## Non-technical explanation
It allows to use SQLAlchemy with OAuth2Authentication.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( x ) Release notes are required, with the following suggested text:

```markdown
* Add support to 'externalAuthentication' SQLAlchemy URL parameter.  ({issue}`#343`)
```

closes #343

### CODE REVIEW

**Comment by IceS2** [comment]
CLA should be good to go (=

File: trino/sqlalchemy/dialect.py [hunk 1]
```diff
@@ -106,6 +111,10 @@ def create_connect_args(self, url: URL) -> Tuple[Sequence[Any], Mapping[str, Any
             kwargs["http_scheme"] = "https"
             kwargs["auth"] = CertificateAuthentication(unquote_plus(url.query['cert']), unquote_plus(url.query['key']))
 
+        if "externalAuthentication" in url.query:
```
[the remark this answers is not in the document]
Comment (IceS2) [reply trino/sqlalchemy/dialect.py:114]: That's what I've been doing when using the client directly. But I've been running into an issue passing it as a class when trying to use it from some other projects such as [ipython-sql](https://github.com/catherinedevlin/ipython-sql).
Eventually I found out that `ipython-sql` allows you to actually pass the `-creator` parameter to bypass the configuration by sending a connection straight away but it'd be great if it would work without this workaround.

File: trino/sqlalchemy/dialect.py [hunk 2]
```diff
@@ -106,6 +111,10 @@ def create_connect_args(self, url: URL) -> Tuple[Sequence[Any], Mapping[str, Any
...
+            kwargs["http_scheme"] = "https"
```
Comment (hashhar) [inline trino/sqlalchemy/dialect.py:115]: why automatically set this? Someone can have configured server to accept credentials over http (it's not recommended but it'll break someone).

File: trino/sqlalchemy/dialect.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hashhar) [reply trino/sqlalchemy/dialect.py:114]: Do we have good behaviour when someone has externalAuthentication in URI and has explicitly set a different auth mechanism? What takes precedence? I'd like to avoid this situation of having multiple ways of configuring something since it leads to more combination of things to test and edge cases to think about.

File: trino/sqlalchemy/dialect.py (hunk 2 above)
Comment (hashhar) [reply trino/sqlalchemy/dialect.py:115]: ingore

File: trino/sqlalchemy/dialect.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (IceS2) [reply trino/sqlalchemy/dialect.py:114]: Hey @hashhar, I have similar feelings to be honest. I just extended the way it's being done for SQLAlchemy URIs (It's already done this way for Basic Authentication, JWT Authentication and Cert Authentication).

I think it makes sense to align with the JDBC driver and have this for SQLAlchemy... If the user passes multiple Auth Methods I think it's on the user side to be honest, but I guess it could be improved by validating if they did this somehow (checking the query args and `connect_args`)

[the remark this answers is not in the document]
Comment (hovaesco) [reply trino/sqlalchemy/dialect.py:114]: This validation seems useful but out of scope of this PR because it needs to apply to other auth methods as outlined above.

**Review by hovaesco — APPROVED** [review]

File: trino/sqlalchemy/dialect.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (IceS2) [reply trino/sqlalchemy/dialect.py:114]: You mean validating if the user passes multiple auth methods?

So you think it'd be fine to keep that responsibility on the user side so far? I'd be fine with it (=

**Comment by hashhar** [comment]
@cla-bot check

**Review by hashhar — APPROVED** [review]

**Comment by esemeniuc** [comment]
merge?

**Comment by hashhar** [comment]
Sorry for forgetting to come back to this. I'll just squash the commits and then merge this.

Thanks for your work @esemeniuc.

**Comment by hashhar** [comment]
Squashed the commits + reworded commit msg, will merge once CI is finished.

**Comment by hashhar** [comment]
Sorry this ended up waiting so long @IceS2.

We plan a release today/tomorrow and will announce on the #python-client on the Trino Slack.

**Comment by IceS2** [comment]
> Sorry this ended up waiting so long @IceS2.
> 
> We plan a release today/tomorrow and will announce on the #python-client on the Trino Slack.

Hey! Don't worry (=
I know sometimes it takes time! I'm glad I could collaborate a bit with you folks o/
